adhocracy4/emails/tasks.py

```diff
@@ -1,16 +1,48 @@
 import importlib
+import logging
+from typing import Optional
 
+from celery import Task
 from celery import shared_task
 from django.apps import apps
 
+log = logging.getLogger(__name__)
 
-@shared_task
+
+@shared_task(bind=True)
 def send_async(
-    email_module_name, email_class_name, app_label, model_name, object_pk, args, kwargs
+    self: Task,
+    email_module_name: str,
+    email_class_name: str,
+    app_label: str,
+    model_name: str,
+    object_pk: Optional[int] = None,
+    args: Optional[list] = None,
+    kwargs: Optional[dict] = None,
+    data: Optional[dict] = None,
 ):
-    imported_module = importlib.import_module(email_module_name)
-    cls = getattr(imported_module, email_class_name)
-    model = apps.get_model(app_label, model_name)
-    obj = model.objects.get(pk=object_pk)
+    task_id = self.request.id[:5]
+
+    if not (data or object_pk):
+        log.error(f"called send_async without object_pk or data: {task_id=}")
+        return
+
+    args = args or []
+    kwargs = kwargs or {}
+
+    log.info(f"{task_id=}, {email_module_name=}, {email_class_name=}")
+
+    email_module = importlib.import_module(email_module_name)
+    email_class = getattr(email_module, email_class_name)
+
+    if data:
+        log.info(f"{task_id=}, {data=}")
+
+        email_class().dispatch(data, *args, **kwargs)
```

Comment:
here either **kwargs or **data, no need for both. Depending which one name you chose, you may need to update the if statements above.

Reply:
we agreed to do any real refactoring in a separate story and focus on making a4 work with celery in this PR.